src/main/scala/cromwell/util/TryUtil.scala

```diff
@@ -16,4 +19,54 @@ object TryUtil {
 
   def stringifyFailures[T](possibleFailures: Traversable[Try[T]]): Traversable[String] =
     possibleFailures.collect { case failure: Failure[T] => stringifyFailure(failure) }
+
+  private def defaultSuccessFunction(a: Any): Boolean = true
+
+  /**
+   * Runs a block of code (`fn`) `retries` number of times until it succeeds.
+   * It will wait `pollingInterval` amount of time between retry attempts and
+   * The `pollingBackOffFactor` is for exponentially backing off the `pollingInterval`
+   * on subsequent retries.  The `pollingInterval` shall not exceed `maxPollingInterval`
+   *
+   * Returns a Try[T] where T is the return value of `fn`, the function to be retried.
+   * If the return value is Success[T] then at least one retry succeeded.
+   *
+   * The isSuccess function is optional but if provided, then isSuccess(fn) must be true
+   * or it will trigger another retry.  if isSuccess is omitted, the only way the fn can
+   * fail is if it throws an exception.
+   *
+   * Use `retries` value of -1 to indicate retrying indefinitely.
+   */
+  @annotation.tailrec
+  def retryBlock[T](fn: Option[T] => T,
+                    isSuccess: T => Boolean = defaultSuccessFunction _,
+                    retries: Int,
+                    pollingInterval: Duration,
+                    pollingBackOffFactor: Double,
+                    maxPollingInterval: Duration,
+                    failMessage: Option[String] = None,
+                    priorValue: Option[T] = None): Try[T] = {
+    Try { fn(priorValue) } match {
+      case Success(x) if isSuccess(x) => Success(x)
+      case value if retries > 1 || retries < 0 =>
+
+        val retryCountMessage = if (retries > 0) s" ($retries more retries) " else ""
+        val retryMessage = s"Retrying in $pollingInterval$retryCountMessage..."
+        if (failMessage.isDefined) logger.warn(failMessage.map(m => s"$m.  $retryMessage").get)
+
+        Thread.sleep(pollingInterval.toMillis)
+
+        retryBlock(
+          fn,
+          isSuccess,
+          retries - 1,
+          Duration(Math.min((pollingInterval.toMillis * pollingBackOffFactor).toLong, maxPollingInterval.toMillis), "milliseconds"),
+          pollingBackOffFactor,
+          maxPollingInterval,
+          failMessage,
+          value.toOption
+        )
+      case f => f
+    }
+  }
```

Comment:
sorry I don't understand what "ignore this" means.  Are you changing this
code in your branch too?

On Thu, Oct 8, 2015 at 9:13 PM, Jeff Gentry <notifications@github.com>
wrote:

> In src/main/scala/cromwell/util/TryUtil.scala
> <https://github.com/broadinstitute/cromwell/pull/227#discussion_r41588940>
> :
>
> > +
> > +        Thread.sleep(pollingInterval.toMillis)
> > +
> > +        retryBlock(
> > +          fn,
> > +          isSuccess,
> > +          retries - 1,
> > +          Duration(Math.min((pollingInterval.toMillis * pollingBackOffFactor).toLong, maxPollingInterval.toMillis), "milliseconds"),
> > +          pollingBackOffFactor,
> > +          maxPollingInterval,
> > +          failMessage,
> > +          value.toOption
> > +        )
> > +      case f => f
> > +    }
> > +  }
>
> Ignore this as my branch is touching on this and I quit touching it (as f
> I get to code anymore) until you merge
> … <#1504a295176b0b27_>
> On Oct 8, 2015, at 8:46 PM, Scott Frazer <notifications@github.com>
> wrote: In src/main/scala/cromwell/util/TryUtil.scala: > + > +
> Thread.sleep(pollingInterval.toMillis) > + > + retryBlock( > + fn, > +
> isSuccess, > + retries - 1, > + Duration(Math.min((pollingInterval.toMillis
> * pollingBackOffFactor).toLong, maxPollingInterval.toMillis),
> "milliseconds"), > + pollingBackOffFactor, > + maxPollingInterval, > +
> failMessage, > + value.toOption > + ) > + case f => f > + } > + } I'm
> certainly not opposed to it but we might lose some of the subtleties like
> optionally printing a warning message to the log on retry, which would be
> useful to know, but we don't want the log message in some cases. We might
> also lose the ability to pass prior attempt's values to future attempts for
> logging purposes (e.g. in JES when we say "Status change from Initializing
> to Running"). I figured the algorithm was straightforward enough to not
> have to add another dependency (I don't know if we already include this as
> a dependency or not). Are you happy with this solution or should I put in
> more effort to look at the alternative? — Reply to this email directly or
> view it on GitHub.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/cromwell/pull/227/files#r41588940>.
>